students/cbrown/Lesson 6/mailroom_part_four.py

```diff
@@ -0,0 +1,170 @@
+          print('\n')
+
+    #not found in list. Add new donor entry
+    else:
+        new_entry = []
+        new_entry.append(amount)
+        donors[donor_name] = new_entry
+
+    return donors
+
+def thank_you_note(donor_name,amount):
+
+    thanks_dict = {'donor_name':donor_name,'amount':amount}
+
+    #Formatted Thank You Note
+    message = ('Dear {donor_name}, \n\nThank you for your show of support and generosity. '
+      'Your Donation of ${amount} will contribute to saving Olympic Marmots '
+      'in Washington State. These Marmota are special and a unique gift to the Olympic '
+      'National Park ecosystem. As a way of saying thank you. '
+      'You will be receiving your very own Olympic Marmot t-shirt in the mail!\n\n'
+      'Sincerely,\n\nThe Olympic Marmot Wildlife Foundation\n').format(**thanks_dict)
+
+    return message
+
+def send_thanks():
+    print('\n',"For A Complete Donor List Type 'List'\n ")
+
+    #Create a unique list of donor names
+    donor_list = [names for names in donors.keys()]
+
+    while True:
+        donor_name = input("What donor(s) are you looking for?\n ").title()
+        if donor_name == 'List':
+            print('\n',"Here is A Complete List of Donor Names\n ")
+            print(donor_list)
+        else:
+            break
+
+    #prompt for donation amount
+    try:
+        amount = float(input("How Much Did This Person Donate?\n "))
+    except ValueError: #start prompt over again if exception occurs
+        print('Please input the amount donated')
+        send_thanks()
+
+    #updates donor database by checking existing donors and adding new ones if they don't exist
+    donor_verify(donor_name,amount,donor_list)
+
+    #formats message thanking the individual donor
+    message = thank_you_note(donor_name,amount)
+
+    print(message)
+
+#function for sorting key
+def sum_value(donations):
+    item = donations[1]
+    return sum(item)
+
+#create a sorted donations dict by total amount
+sorted_donations = sorted(donors.items(),key = sum_value,reverse=True)
+
+def report_rows(name,amounts):
+    total_given = sum(amounts)
+    num_gift = len(amounts)
+    avg_gift = total_given/num_gift
+
+    total_given = '{:,.0f}'.format(total_given)
+    avg_gift = '{:,.0f}'.format(avg_gift)
+    row = '{:<25}${:^15}{:^15}${:^15}'.format(name,total_given,num_gift,avg_gift)
+
+    return row
+
+def print_report(report_lines):
+    '''
+    Prints the output of the donor report using data generated from report_rows function & report_lines list
+    '''
+    #prints the header of the report
+    print('{:<25}{}{:^15}{}{:^15}{}{:^15}'.format('Donor Name','|','Total Given','|','Num Gifts','|','Average Gift'))
+    str_len = len('{:<25}{}{:^15}{}{:^15}{}{:^15}'.format('Donor Name','|','Total Given','|','Num Gifts','|','Average Gift'))
+    print('-' * str_len)
+
+    #prints name, total, num of gifts, and avg gift
+    for row in report_lines:
+        print(row)
```

Comment:
Report is not picking up new donors and donations

Reply:
Fixed this and resubmitted!